### PR TITLE
Read video from memory newapi

### PR DESCRIPTION
**Pull request by jdsgomes**
adds read from memory functionality in the new video API.

Addresses #6603

### CODE REVIEW

File: torchvision/io/video_reader.py
Comment (jdsgomes) [inline torchvision/io/video_reader.py:75]: Possibly it is a bad idea to change this, despite being a beta API, but leaving it as path because misleading too.. I am open for suggestions from the reviewers

Comment (YosuaMichael) [reply torchvision/io/video_reader.py:75]: Since our Beta API has no warning (which make few users aware of it), I think we should follow deprecation policy for BC changes. I suggest we follow similar approach to transform to deprecate `resample` param: https://github.com/pytorch/vision/blob/21790df9a4f77bd9ec4db44de04594cb539457a7/torchvision/transforms/transforms.py#L1275-L1278

Basically we allow both `path` and `src` (put `path` as optional and the last param), and give warning for those who use `path` that it will be removed on v 0.17.

Comment (jdsgomes) [reply torchvision/io/video_reader.py:75]: But in that case if someone is using  `VideoReader(path="bla.avi")` wont be able to do it. So I am thinking that we should do the other way around, i.e., we add `src` as optional for now and in two releases it replaces `path`. lmk your thoughts

Comment (jdsgomes) [reply torchvision/io/video_reader.py]: Actually I don't see any good option

**Option 1: Make path optional**
This is non BC because you no longer can do `VideoReader(path="bla.avi")`. 

**Option 2: Make src optional**
This option is BC and allows users to do `VideoReader(path="bla.avi")`, `VideoReader("bla.avi")` and `VideoReader(bytesobj)` but if one wants to use the src keyword then it looks weird: `VideoReader("", src="bla.avi")`. Furthermore later wen we deprecate path it will be a BC change



**Option 3: add an extra argument permanently**
Third option  will be to add `video_obj` as last optional argument, and when this was provided then we we would ignore the `path`. This is BC but a bit ugly when we want to construct a VideoReader from memory : `VideoReader("", video_obj=bytesobj)`

Given this I am inclined for option 3, unless I am missing something

Comment (YosuaMichael) [reply torchvision/io/video_reader.py]: My thinking is similar with Option 2, but we let path to be optional at the end, and we set default value of src as empty string.
```
VideoReader(src: Any = "", ..., path: Optional[str] = None)
```

and inside constructor we can handle like:
```
if src == "":
  if path is None:
    raise ValueError(...)
  src = path
```

what do you think?

Comment (jdsgomes) [reply torchvision/io/video_reader.py:75]: yeah that makes a lot of  sense.  Thanks

**Comment by vedantroy** [comment]
This segfaults on certain videos. Here's a partial-reproduction script (the videos are stored in a pandas dataframe):

```
import pandas as pd

# df2 = pd.read_pickle("df2.pkl")
df = pd.read_pickle("df.pkl")
# print the # of rows in the df
print(len(df))
# print the keys in the df
print(df.keys())
# print the first row in the df
print(df.iloc[0])


# print the type of the 1st value in the 1st row
first_vid = df.iloc[0][0]
print(f"Length: {len(first_vid)}")
print(f"Type: {type(first_vid)}")

# write first_vid to a file
with open("test.mp4", "wb") as f:
    f.write(first_vid)

import itertools
import copy

import torch
from torchvision.io import VideoReader
import torchvision

def clip_from_start(buf: bytes, expected_frames: int):
    # import av
    # import io
    # buffer = io.BytesIO(buf)
    # container = av.open(buffer)
    # i = 0 
    # for frame in container.decode(video=0):
    #     print(type(frame))
    #     i += 1
    #     print(i)
    #     pass

    tensor = torch.frombuffer(buf, dtype=torch.uint8)
    tensor = copy.deepcopy(tensor)
    # torchvision.io.read_video()
    rdr = VideoReader(tensor)
    sampled_frames = list(itertools.islice(iter(rdr), expected_frames))
    if len(sampled_frames) != expected_frames:
        return None
    data = []
    for frame in sampled_frames:
        data.append(frame["data"])
    return torch.stack(data, dim=0)

clip = clip_from_start(first_vid, 2)
print(clip.shape)
```

I'm working to get approval of the public copy of the data.

File: torchvision/io/video_reader.py
Comment (YosuaMichael) [inline torchvision/io/video_reader.py:130]: NIT: I just realized that missing argument should raise `TypeError` instead of `ValueError`. Can we change this to `TypeError` instead?

Comment (jdsgomes) [reply torchvision/io/video_reader.py:130]: In this case I disagree, we are checking for the value of the variable, not for the type. Would like to understand your interpretation in case I am missing something.

Comment (YosuaMichael) [reply torchvision/io/video_reader.py:130]: I just tried on my local python the following:
```
>>> def foo(x, y):
...  print(x, y)
... 
>>> foo("hello")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() missing 1 required positional argument: 'y'
```

Notice that when we dont specify 1 required positional argument, python raise `TypeError` instead of `ValueError` (I actually surprised that python use `TypeError` for this). Therefore I think we should follow this convention since in this case we basically missing 1 required argument as well.

**Comment by jdsgomes** [comment]
> This segfaults on certain videos. Here's a partial-reproduction script (the videos are stored in a pandas dataframe):
> 
> ```
> import pandas as pd
> 
> # df2 = pd.read_pickle("df2.pkl")
> df = pd.read_pickle("df.pkl")
> # print the # of rows in the df
> print(len(df))
> # print the keys in the df
> print(df.keys())
> # print the first row in the df
> print(df.iloc[0])
> 
> 
> # print the type of the 1st value in the 1st row
> first_vid = df.iloc[0][0]
> print(f"Length: {len(first_vid)}")
> print(f"Type: {type(first_vid)}")
> 
> # write first_vid to a file
> with open("test.mp4", "wb") as f:
>     f.write(first_vid)
> 
> import itertools
> import copy
> 
> import torch
> from torchvision.io import VideoReader
> import torchvision
> 
> def clip_from_start(buf: bytes, expected_frames: int):
>     # import av
>     # import io
>     # buffer = io.BytesIO(buf)
>     # container = av.open(buffer)
>     # i = 0 
>     # for frame in container.decode(video=0):
>     #     print(type(frame))
>     #     i += 1
>     #     print(i)
>     #     pass
> 
>     tensor = torch.frombuffer(buf, dtype=torch.uint8)
>     tensor = copy.deepcopy(tensor)
>     # torchvision.io.read_video()
>     rdr = VideoReader(tensor)
>     sampled_frames = list(itertools.islice(iter(rdr), expected_frames))
>     if len(sampled_frames) != expected_frames:
>         return None
>     data = []
>     for frame in sampled_frames:
>         data.append(frame["data"])
>     return torch.stack(data, dim=0)
> 
> clip = clip_from_start(first_vid, 2)
> print(clip.shape)
> ```
> 
> I'm working to get approval of the public copy of the data.
> 
> In the meantime, the error is:
> 
> ```
> test.py:41: UserWarning: The given buffer is not writable, and PyTorch does not support non-writable tensors. This means you can write to the underlying (supposedly non-writable) buffer using the tensor. You may want to copy the buffer to protect its data or make it writable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at ../torch/csrc/utils/tensor_new.cpp:1563.)
>   tensor = torch.frombuffer(buf, dtype=torch.uint8)
> malloc(): corrupted top size
> Aborted (core dumped)
> ```

Thanks for reporting this. 
I cannot reproduce the same error with some test videos. Is there any possibility that the data is corrupt? 
Otherwise please let me know when you would be able to share the sample video for us to verify and debug. 

Thanks

File: test/test_videoapi.py
Comment (bjuncek) [inline test/test_videoapi.py:147]: Given that this is like-for-like decoding (using verifiably the same backend), we should probably insist on either a max delta or completely equal.

File: torchvision/csrc/io/decoder/defs.h
Comment (bjuncek) [inline torchvision/csrc/io/decoder/defs.h:168]: Why this? 
(don't know C++ well, so I'm just curious)

**Review by bjuncek — CHANGES_REQUESTED** [review]
Looks good to me. I'd love to make the tests stricter but otherwise good to go :)

File: torchvision/csrc/io/decoder/defs.h
Comment (jdsgomes) [reply torchvision/csrc/io/decoder/defs.h:168]: just want to initialise  the uri value as empty string

File: test/test_videoapi.py
Comment (jdsgomes) [reply test/test_videoapi.py:141]: actually there are two videos for which this is not true: `SOX5yA1l24A.mp4`
I verified with this code by reading twice from file and also the result is not the same:

```python
In [3]: reader1 = VideoReader('test/assets/videos/SOX5yA1l24A.mp4')
In [4]: reader2 = VideoReader('test/assets/videos/SOX5yA1l24A.mp4')
In [5]: reader1_frames = []
In [6]: reader2_frames = []

In [8]: for r1frame in reader1:
   ...:     reader1_frames.append(r1frame["data"])

In [9]: for r2frame in reader2:
   ...:     reader2_frames.append(r2frame["data"])

In [12]: for i in range(len(reader1_frames)):
    ...:     if torch.all(reader1_frames[i].eq(reader2_frames[i])):
    ...:         print(f"Frame {i} is not equal")

Frame 0 is not equal
Frame 2 is not equal
Frame 3 is not equal
Frame 4 is not equal
Frame 5 is not equal
Frame 6 is not equal
Frame 7 is not equal
Frame 8 is not equal
Frame 9 is not equal
Frame 10 is not equal
Frame 11 is not equal
Frame 12 is not equal
....
```

Do you know if for this particular format the decoding could be non deterministic?

File: test/test_videoapi.py
Comment (jdsgomes) [inline test/test_videoapi.py]: This was added in the wrong section so reverted the commit. My comment regarding why we cannot make the tests strict from read_from_memory still holds

File: test/test_videoapi.py
Comment (YosuaMichael) [reply test/test_videoapi.py:147]: This is actually surprising to me, is this non deterministic behaviour also occur on the stable API (if we specify same pts)?

Comment (bjuncek) [reply test/test_videoapi.py:147]: Yup; we use the same backend.
@jdsgomes can you check if it happens on the `pyav`?

Maybe this is a cue that we should reconsider our selection of testing videos.
Specifically, I think we should re-sample them again from whatever are most used datasets these days, plus some FFMPEG fate situations.

Comment (jdsgomes) [reply test/test_videoapi.py:147]: I confirmed that that is the case.

Comment (jdsgomes) [reply test/test_videoapi.py:147]: please note that there is a bug in the test code (checking for equal instead of not equal) but still there are some frames that are different. This happens both in pyav and video_reader

**Review by bjuncek — APPROVED** [review]
Approving to unblock, and starting a new issue to track video assets.

**Pull request merged**
